### PR TITLE
SONAR-9616 Ability to set the name of the main branch

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

**Comment by julienlancelot** [comment]
Compilation is failing because of missing file header in RenameActionTest.java

File: server/sonar-server/src/test/java/org/sonar/server/projectbranch/ws/RenameActionTest.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    userSession.logIn();
+
+    expectedException.expect(IllegalArgumentException.class);
+    expectedException.expectMessage("The 'branch' parameter is missing");
+
+    tester.newRequest().setParam("project", "projectName").execute();
+  }
+
+  @Test
+  public void fail_if_not_logged_in() {
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/projectbranch/ws/RenameActionTest.java]: No need to check if user if logged or not, but please add a unit test to check that the user has the correct permission

**Comment by julienlancelot** [comment]
Doing some functional tests, it's not possible to rename a master branch twice using the same name => it should not fail if you try to rename the main branch with the same name

File: server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    action
+      .createParam(PARAM_BRANCH)
+      .setDescription("New name of the main branch")
+      .setExampleValue("master")
+      .setRequired(true);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    userSession.checkLoggedIn();
```
Comment (dbmeneses) [inline server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java]: it's a fail fast check, before querying the project

File: server/sonar-db-dao/src/main/java/org/sonar/db/component/BranchDao.java [hunk 3]
```diff
@@ -49,6 +49,15 @@ public void upsert(DbSession dbSession, BranchDto dto) {
     }
   }
 
+  public void updateMainBranchName(DbSession dbSession, String projectUuid, String newBranchKey) {
```
Comment (dbmeneses) [inline server/sonar-db-dao/src/main/java/org/sonar/db/component/BranchDao.java]: same thing is done in ComponentKeyUpdaterDao. I can move the check to the WS but IMO upsert should not update the key. Generic methods to update a row should not change the ID of the row (and currently it doesn't). It's better to have a dedicated method for it.

File: server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    String projectKey = request.mandatoryParam(PARAM_PROJECT);
+    String branchKey = request.mandatoryParam(PARAM_BRANCH);
+
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      ComponentDto project = componentFinder.getRootComponentByUuidOrKey(dbSession, null, projectKey);
+      checkPermission(project);
+
+      BranchDto branch = checkFoundWithOptional(dbClient.branchDao().selectByUuid(dbSession, project.uuid()), "Project '%s' not found", projectKey);
```
Comment (dbmeneses) [inline server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java]: ok

File: server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      dbClient.branchDao().updateMainBranchName(dbSession, branch.getUuid(), branchKey);
+      dbSession.commit();
+      response.noContent();
+    }
+  }
+
+  private void checkPermission(ComponentDto project) {
+    userSession.hasComponentPermission(UserRole.ADMIN, project);
```
Comment (dbmeneses) [inline server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java]: good catch. Did the same mistake for DeleteAction

File: server/sonar-db-dao/src/main/java/org/sonar/db/component/BranchDao.java (hunk 3 above)
Comment (julienlancelot) [inline server/sonar-db-dao/src/main/java/org/sonar/db/component/BranchDao.java]: ComponentKeyUpdaterDao is a bad example as it should be refactored :(.
If you don't want to use upsert, then create an update method

Comment (dbmeneses) [inline server/sonar-db-dao/src/main/java/org/sonar/db/component/BranchDao.java]: I think that any generic update method should not change the key. Also, it doesn't make much sense to have to give an entire branchDto to change the key. There should be a updateMainBranchName to change the row's key by querying the projectUuid. This makes the change atomic and makes sure that no other field that would be included in the BranchDto gets changed at the same time.

File: server/sonar-db-dao/src/main/resources/org/sonar/db/component/BranchMapper.xml [hunk 6]
```diff
@@ -35,6 +35,14 @@
       #{now, jdbcType=BIGINT}
     )
   </insert>
+  
+  <update id="updateMainBranchName" parameterType="map">
+    update project_branches
+    set
+      kee = #{newBranchName, jdbcType=VARCHAR}
+    where
+      uuid = #{projectUuid, jdbcType=VARCHAR}
```
Comment (julienlancelot) [inline server/sonar-db-dao/src/main/resources/org/sonar/db/component/BranchMapper.xml]: updated_at needs to be refresh to current time

File: server/sonar-db-dao/src/test/java/org/sonar/db/component/BranchDaoTest.java [hunk 7]
```diff
@@ -43,6 +44,9 @@
   private System2 system2 = new TestSystem2().setNow(NOW);
 
   @Rule
+  public ExpectedException exception = ExpectedException.none();
```
Comment (julienlancelot) [inline server/sonar-db-dao/src/test/java/org/sonar/db/component/BranchDaoTest.java]: No more needed :)

File: server/sonar-db-dao/src/test/java/org/sonar/db/component/BranchDaoTest.java [hunk 8]
```diff
@@ -87,11 +117,11 @@ public void insert_branch_with_all_fields_and_max_length_values() {
     underTest.insert(dbSession, dto);
 
     Map<String, Object> map = db.selectFirst(dbSession, SELECT_FROM + " where uuid='" + dto.getUuid() + "'");
```
Comment (julienlancelot) [inline server/sonar-db-dao/src/test/java/org/sonar/db/component/BranchDaoTest.java]: Minor - BranchDao#selectByUuid could have been used instead

File: server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    this.dbClient = dbClient;
+    this.componentFinder = componentFinder;
+    this.userSession = userSession;
+  }
+
+  @Override
+  public void define(NewController context) {
+    WebService.NewAction action = context.createAction(ACTION_RENAME)
+      .setSince("6.6")
+      .setDescription("Rename the main branch of a project. Requires 'Administer' permission on the specified project.")
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java]: Please add a <br/> between the 2 sentences.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [DeleteAction.java#L41](https://github.com/SonarSource/sonarqube/blob/fcea7a6fefdf4b58536d651befdc0881f8fa47b1/server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/DeleteAction.java#L41): Remove this empty statement. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AEmptyStatementUsageCheck)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    userSession.checkLoggedIn();
+    String projectKey = request.mandatoryParam(PARAM_PROJECT);
+    String branchKey = request.mandatoryParam(PARAM_BRANCH);
+
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      ComponentDto project = componentFinder.getRootComponentByUuidOrKey(dbSession, null, projectKey);
+      checkPermission(project);
+
+      Optional<BranchDto> branch = dbClient.branchDao().selectByKey(dbSession, project.uuid(), BranchKeyType.BRANCH, branchKey);
+      if (branch.isPresent()) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java]: You need to check that it's not the current branch => You need to load the BranchDto from the main branch before, and verify it's not the same uuid as the one you've found

File: server/sonar-server/src/test/java/org/sonar/server/projectbranch/ws/RenameActionTest.java [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    expectedException.expectMessage("Insufficient privileges");
+
+    tester.newRequest()
+      .setParam("project", project.getKey())
+      .setParam("branch", "branch1")
+      .execute();
+  }
+
+  @Test
+  public void successfully_rename() {
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/projectbranch/ws/RenameActionTest.java]: Add a UT to test renaming of the main branch with the same name => it should not fail

File: server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java (hunk 9 above)
Comment (dbmeneses) [inline server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java]: I guess you wrote  `<br />`

File: server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java (hunk 10 above)
Comment (dbmeneses) [inline server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java]: Good point. But it should be enough to change the condition to `branch.isPresent() && !branch.isMain()` (will add tests to confirm)

File: server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java (hunk 9 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java]: ohoh, indeed !

File: server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java (hunk 10 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/RenameAction.java]: Indeed !

**Pull request closed without merging**